test/test_main_type_hint.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2024 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Validate the type signatures on ops.main().
+
+This file doesn't contain any run-time tests, rather we rely on pyright to run over this code.
+Assignment to a variable declared to follow a protocol is equivalent to backwards compatibility.
+"""
+
+from dataclasses import dataclass
+from typing import Optional, Protocol, Type
+
+import ops
+
+
+class CallableWithCharmClassOnly(Protocol):
+    """Encapsulate main function type for simple charms.
+
+    Supports:
+    - ops.main(SomeCharm)
+    - ops.main(charm_class=SomeCharm)
+    """
+
+    def __call__(self, charm_class: Type[ops.charm.CharmBase]): ...
+
+
+class CallableWithCharmClassAndStorageFlag(Protocol):
+    """Encapsulate main function type for advanced charms.
+
+    Supports permutations of:
+    - ops.main(SomeCharm, False)
+    - ops.main(charm_class=SomeCharm, use_juju_for_storage=False)
+    """
+
+    def __call__(
+        self, charm_class: Type[ops.charm.CharmBase], use_juju_for_storage: Optional[bool] = None
+    ): ...
+
+
+class CallableWithoutArguments(Protocol):
```

Comment:
Is this needed (`CallableWithoutArguments`)? Likewise, is `CallableWithCharmClassOnly` necessary? If a value conforms to `CallableWithCharmClassAndStorageFlag`, then (a) it can't be called without arguments and (b) it can be called with the first argument only (since the second argument has a default value).

Perhaps this could be reduced to:
```
class MainCall(Protocol):
    def __call__(
        self, charm_class: Type[ops.charm.CharmBase], use_juju_for_storage: Optional[bool] = None
    ): ...
```

The assignments to type check `main` can then be reduced to
```
def top_level_import() -> None:
    """Recommended style."""
    import ops

    main1: MainCall = ops.main
    main2: MainCall = ops.main.main   # deprecated style

def submodule_import() -> None:
    """Deprecated style."""
    import ops.main

    main1: MainCall = ops.main # type: ignore[assignment]
    # type checker limitation https://github.com/microsoft/pyright/issues/8830
    main2: MainCall = ops.main.main

def import_from_top_level_module() -> None:
    """Also ok."""
    from ops import main

    main1: MainCall = main
    main2: MainCall = main.main  # deprecated style


def import_from_submodule() -> None:
    """Deprecated style."""
    from ops.main import main

    main1: MainCall = main
```

Even if both `CallableWithCharmClassOnly` and `CallableWithCharmClassAndStorageFlag` are necessary, I'd favour removing `CallableWithoutArguments` since there isn't really a nice way do something like `with raises(AssignmentError)` for type checking afaik, so those lines are currently just `type: ignore`d anyway.

Reply:
CallableWithCharmClassOnly is the main way the charms are written.

The two-argument variety is rare, and may go away one day, as Juju storage will no longer be supported in Juju 4.

The point of WithoutArguments is to two-fold, and one could argue that it could be dropped from the test:
- as a negative test (if charm code looks like this, pyright will error out, helping charmers write code that actually works)
- as a sort-of self-test (as in to validate the entire approach of type testing)

I think that overall I went for "explicit is better than implicit".

I think that only testing against CallableWithCharmClassAndStorageFlag has this downside: imagine that one day, someone merges this change:

```diff
 class _Main:
     def __call__(
-        self, charm_class: Type[charm.CharmBase], use_juju_for_storage: Optional[bool] = None
+        self, charm_class: Optional[Type[charm.CharmBase]] = None, use_juju_for_storage: Optional[bool] = None
     ):
         return _main.main(charm_class=charm_class, use_juju_for_storage=use_juju_for_storage)
```

Let's assume that we want to catch that (a separate debate), then with this PR, pyright will produce errors in 2 places:
- test_main_invocation `ops.main() # type: ignore`
- test_main_type_hint `bad = ops.main # type: ignore`

On one hand, type tests help us prevent regressions.

On the other, they are somewhat redundant.

Reply:
That makes sense. I'd forgotten about the ability of the type checker to warn on unneeded `type: ignore`s, in light of which I definitely think it's worth keeping the `CallableWithoutArguments` version for the 'self-test' reason.

I'm not sure I see how the 'negative' reason works though. Won't pyright complain to charm authors the same way about improperly calling main regardless of whether any of these tests exist or not?

I still don't fully understand the benefit of `CallableWithCharmClassOnly` either. Even if most charms are called this way, main is currently only implemented as `CallableWithCharmClassAndStorageFlag`. I guess having both tests means that if the storage flag is eventually removed, then the `CallableWithCharmClassOnly` tests are already implemented and in place, so it's just a matter of removing the old storage flag tests rather than adding new ones. Are there any other benefits I'm missing here?

In any case, don't let me hold you up with my questions, I suspect I just haven't fully gotten my head around the benefit of these tests.

Reply:
Following conversation in daily sync, I'm happy to resolve this. For anyone reading, I think I'm simply questioning whether some of these checks are redundant, and the resolution is that some may be, but it's worth having them for safety, especially since ideally these tests can all be removed eventually anyway if legacy main goes away.